src/Strings.jl

```diff
@@ -1,9 +1,98 @@
 
 using ..IOExtras
 
+# A `Base.VersionNumber` is a SemVer spec, whereas a HTTP versions is just 2 digits,
+# This allows us to use a smaller type and more importantly write a simple parse method
+# that avoid allocations.
+"""
+    HTTPVersion(major, minor)
+
+The HTTP version number consists of two digits separated by a
+"." (period or decimal point). The first digit (`major` version)
+indicates the HTTP messaging syntax, whereas the second digit (`minor`
+version) indicates the highest minor version within that major
+version to which the sender is conformant and able to understand for
+future communication.
+
+See [RFC7230 2.6](https://tools.ietf.org/html/rfc7230#section-2.6)
+"""
+struct HTTPVersion
+    major::UInt8
+    minor::UInt8
+end
+
+HTTPVersion(major::Integer) = HTTPVersion(major, 0x00)
+HTTPVersion(v::AbstractString) = parse(HTTPVersion, v)
+HTTPVersion(v::VersionNumber) = convert(HTTPVersion, v)
+# Lossy conversion. We ignore patch/prerelease/build parts even if non-zero/non-empty,
+# because we don't want to add overhead for a case that should never be relevant.
+Base.convert(::Type{HTTPVersion}, v::VersionNumber) = HTTPVersion(v.major, v.minor)
+Base.VersionNumber(v::HTTPVersion) = VersionNumber(v.major, v.minor)
+
+Base.show(io::IO, v::HTTPVersion) = print(io, "HTTPVersion(\"", string(v.major), ".", string(v.minor), "\")")
+
+Base.:(==)(va::VersionNumber, vb::HTTPVersion) = va == VersionNumber(vb)
+Base.:(==)(va::HTTPVersion, vb::VersionNumber) = VersionNumber(va) == vb
+
+Base.isless(va::VersionNumber, vb::HTTPVersion) = isless(va, VersionNumber(vb))
+Base.isless(va::HTTPVersion, vb::VersionNumber) = isless(VersionNumber(va), vb)
+function Base.isless(va::HTTPVersion, vb::HTTPVersion)
+    va.major < vb.major && return true
+    va.major > vb.major && return false
+    va.minor < vb.minor && return true
+    return false
+end
+
+function Base.parse(::Type{HTTPVersion}, v::AbstractString)
+    ver = tryparse(HTTPVersion, v)
+    ver === nothing && throw(ArgumentError("invalid HTTP version string: $(repr(v))"))
+    return ver
+end
+
+# We only support single-digits for major and minor versions
+# - we can parse 0.9 but not 0.10
+# - we can parse 9.0 but not 10.0
+function Base.tryparse(::Type{HTTPVersion}, v::AbstractString)
+    isempty(v) && return nothing
+    len = ncodeunits(v)
+
+    i = firstindex(v)
+    d1 = v[i]
+    if isdigit(d1)
+        major = parse(UInt8, d1)
+    else
+        return nothing
+    end
+
+    i = nextind(v, i)
+    i > len && return HTTPVersion(major)
+    dot = v[i]
+    dot == '.' || return nothing
+
+    i = nextind(v, i)
+    i > len && return HTTPVersion(major)
+    d2 = v[i]
+    if isdigit(d2)
+        minor = parse(UInt8, d2)
+    else
+        return nothing
+    end
+    return HTTPVersion(major, minor)
+end
+
+# const RE_HTTP_VERSION = r"(\d+)(?:\.(\d+))?"
+# function Base.tryparse(::Type{HTTPVersion}, v::AbstractString)
+#     m = match(RE_HTTP_VERSION, String(v)::String)
```

Comment:
this is how `VersionNumber` construction works, for comparison

I need to remove this